### ISSUE TITLE
Issue: Tracing fails on larger batches

### ISSUE THREAD

**Issue by jacob-israel-turner** [langchain-ai/langsmith-sdk#1101]
### Issue you'd like to raise.

Hello LangSmith team!

## Description

I'm running a small chain that takes an image and returns a transcription and description of the image. I'm doing a [batch operation](https://js.langchain.com/docs/how_to/lcel_cheatsheet/#batch-a-runnable), and LangSmith fails to save the trace for most of the runs. I get this error on ~50 out of 61 runs on a document I've been testing on:

```
DOMException [TimeoutError]: The operation was aborted due to timeout
    at node:internal/deps/undici/undici:12502:13
    at RetryOperation._fn ([REDACTED]/node_modules/.pnpm/p-retry@4.6.2/node_modules/p-retry/index.js:50:12) {
  attemptNumber: 7,
  retriesLeft: 0
}
```

It seems like the traces should all be batched and reported together, but that's not happening - I'm getting ~50 of these errors logged every time I run this batch. It's a terrible dev experience.

Additionally, the reported latency times in the LangSmith console are incorrect. It looks like it's reporting the diff from chain start time to the time it was received by the LangSmith server (it _should_ be from chain start to resolution, ignoring latency between my server and the LangSmith server). For the handful of successful traces getting through, I'm seeing reported times of over 100 seconds, while I can see from local logging that the entire batch takes between 45 and 65 seconds to run.

It should be noted that I tried wrapping the function containing the `.batch` call as a [traceable](https://docs.smith.langchain.com/old/tracing/faq/logging_and_viewing#using-traceable--traceable) in the hopes of coaxing LangSmith to report the whole trace in one batch, but I got this error:
```
Error: Failed to batch create run. Received status [403]: Forbidden. Server response: <!doctype html><meta charset="utf-8"><meta name=viewport content="width=device-width, initial-scale=1"><title>403</title>403 Forbidden
```
## My Setup

I can confirm my env variables are setup properly, as the automatic tracing functionality just works.  `LANGCHAIN_CALLBACKS_BACKGROUND` is set to true.

LangSmith TypeScript SDK version: `0.1.66`

## Potential solutions

- Ideally, calling `.batch` on a LangChain runnable would cause LangSmith to report a single trace.
- Alternatively, I'd like to be able to modify the timeout on LangSmith to see if I can find a configuration that works
- As a last resort, I need the ability to suppress failed trace reporting. These errors are making my console unusable.

**Comment by jacob-israel-turner** [comment]
My intuition tells me this is probably because of the _size_ of the traces. Each one includes a base64-encoded image. Another solution might be to allow users to disable specific fields from traces?

**Comment by simply10w** [comment]
We think this same issue was causing a very tricky bug on our side, it was either breaking entire tool call, or making the tool take very long time to complete, like we are talking 50 seconds-100seconds for a simple ReAct agent with just RAG.
<img width="1205" alt="Screenshot 2024-10-24 at 17 58 54" src="https://github.com/user-attachments/assets/6180ad3b-62ad-400c-9667-f603ba8d81f5">

It tooks us 2 weeks to find that by removing langsmith tracing from that part of the agent it stopped happening. 

The error looked initially like it was coming from weaviate, and we touched base with their team and after days of back and forth neither them or us were able to figure it out.

So it seems something gets messed up in the tracing and the stack trace also gets mixed up probably as multiple things are running in the background as callbacks.

The crazy thing was it does not happen on our other AI agents on the same application, it does process around 10.000 tokens in average, and whenever we do 2-3 parallel calls is when the issue happens.

**Comment by jacoblee93** [comment]
Sorry for missing this folks - we're working on a new way to upload images/large traces. Will update here when ready.

**Comment by jacoblee93** [comment]
We also have some behavior where the tracer will block when a root run will block on final save to ensure they finish in serverless environments. We will also look into making this not the case.

**Comment by jacoblee93** [comment]
#1106 should alleviate this for now, will release shortly and update this thread

**Comment by jacoblee93** [comment]
Hey folks, waiting on final review but `0.2.0-rc.0` is live with some performance improvements - if you're up for it, please install!

**Comment by simply10w** [comment]
Hey @jacoblee93  i think its not solved yet, we still get random very slow responses from our agents and looking at the traces this is what i see 

<img width="390" alt="Screenshot 2024-10-25 at 21 28 36" src="https://github.com/user-attachments/assets/dcd099fe-bfa5-431a-82d3-174b88f2a77c">
<img width="350" alt="Screenshot 2024-10-25 at 21 29 40" src="https://github.com/user-attachments/assets/e33f2d6d-381b-431b-998c-d0fc133e844e">

**Comment by jacoblee93** [comment]
@simply10w could you share/DM me one of those traces?

At the very least it shouldn't block the response to the user now - are you seeing large end-user latencies?

**Comment by simply10w** [comment]
@jacoblee93  sure, will DM in a moment 

it is actually causing large user end latencies and also causing the whole server to become unresponsive for a moment as well, when we remove langsmith from these agents that are processing larger amounts of tokens it stops

**Comment by jacoblee93** [comment]
Hey folks, worked with @simply10w a bit and I believe that #1133 and version 0.2.2 should contain the fix!

If you all could try bumping to that and letting me know I'd greatly appreciate it.

**Comment by jacob-israel-turner** [comment]
I tested a bit over the weekend, and it does not seem that 0.2.2 fixes the issue. I will try 0.2.3 this week.

**Comment by jacoblee93** [comment]
Thanks - would be happy to hop on and debug with you this week if you DM me @hacubu on X.

**Comment by ViktoriaLoe** [comment]
Are Recent Batch Tracing Fixes Available for LangSmith in Python?
I saw that recent fixes for batch tracing issues and timeout errors (e.g., #1101) were implemented in the npm version of LangSmith. Could you confirm if these fixes are also available in the Python SDK? If so, which version includes them?
Would be great if the batched runs could be traced together.

Thanks for your help!

**Comment by jacoblee93** [comment]
CC @hinthornw

**Comment by hinthornw** [comment]
Hi @ViktoriaLoe we've made a lot of tracing improvements on the python side in recent versions to help us increase reliability and the max permitted payload sizes.

> "Batched runs could be traced together" 

Could you elaborate what you mean by this? We've had batched run ingestion implemented since last spring at least. We recently implemented a multipart endpoint to make it even faster

**Comment by jacob-israel-turner** [comment]
> Thanks - would be happy to hop on and debug with you this week if you DM me @Hacubu on X.
> 
> You should also verify that you don't have any old LangSmith versions lying around in your lockfile

Thanks for your willingness to help @jacoblee93. I realized I actually ended up migrating my code away from `.batch` so I could move forward. I'm not sure when I'll have the time to go back and test `.batch` with the latest version of LangSmith.

**Comment by jacoblee93** [comment]
> Ah really, my abatch of RunnableLambdas are being traced as all separate calls, how would I get them traced in Langsmith together then?

That's intended - you could wrap them in another parent lambda to get them in the same tree though

**Comment by jacob-israel-turner** [comment]
> > Thanks for your willingness to help @jacoblee93. I realized I actually ended up migrating my code away from .batch so I could move forward. I'm not sure when I'll have the time to go back and test .batch with the latest version of LangSmith.
> 
> Ah ok sorry about this @jacob-israel-turner :(
> 
> I think the latest client should always opt-into batch tracing. For larger images - I think this will help:
> 
> #1113

Ooh yes, that looks very helpful. I had to disable tracing altogether today, as we were going over our hourly bandwidth rate limit.

I get the impression that our use-case is quite different from the norm. Most of our calls through LangChain are doing image transcriptions, and we have very bursty workloads - hundreds or thousands of calls within just a few minutes. It might be nice to allow us to disable sending images with the traces altogether - unless #1113 compresses the images enough to eliminate the rate-limit issue.

Thanks for your hard work @jacoblee93!

**Comment by jacob-israel-turner** [comment]
@jacoblee93 I've been doing a little testing today with tracing enabled on our image processing chain. _Without_ batch enabled, we're definitely still seeing some failed traces - ~10-20%.

Is there another ticket tracking failed traces with large images? Would it be helpful for me to open one?

**Comment by jacoblee93** [comment]
Yes please open another one - sorry for the delay in landing #1113
